packages/babel-helpers/src/helpers/applyDecs2305.ts

```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
+        );
+      }
+    }
+
+    var decs = decInfo[0],
+      decVal = decInfo[3],
+      _: any,
+      isClass = !ret;
+
+    if (!isClass) {
+      if (!decoratorsHaveThis && !Array.isArray(decs)) {
+        decs = [decs];
+      }
+
+      var desc: PropertyDescriptor = {},
+        init: Function[] = [],
+        key: "get" | "set" | "value" =
+          kind === PROP_KIND.GETTER
+            ? "get"
+            : kind === PROP_KIND.SETTER || isAccessor
+              ? "set"
+              : "value";
+
+      if (isPrivate) {
+        if (isField || isAccessor) {
+          desc = {
+            get: setFunctionName(
+              function (this: any) {
+                return decVal(this);
+              },
+              name,
+              "get",
+            ),
+            set: function (this: any, value: any) {
+              decInfo[4](this, value);
+            },
+          };
+        } else {
+          desc[key] = decVal;
+        }
+
+        if (!isField) {
+          setFunctionName(
+            desc[key],
+            name,
+            kind === PROP_KIND.METHOD ? "" : key,
+          );
+        }
+      } else if (!isField) {
+        desc = Object.getOwnPropertyDescriptor(Class, name);
+      }
+    }
+
+    var newValue;
+
+    for (var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1) {
+      var dec = (decs as Function[])[i],
+        decThis = decoratorsHaveThis ? (decs as any[])[i - 1] : void 0;
+
+      var decoratorFinishedRef: DecoratorFinishedRef = {};
+      var ctx: DecoratorContext = {
+        kind: [
+          "field",
+          "accessor",
+          "method",
+          "getter",
+          "setter",
+          "field",
+          "class",
+        ][kind] as any,
+
+        name: name,
+        metadata: metadata,
+      };
+
+      ctx.addInitializer = function (
+        decoratorFinishedRef: DecoratorFinishedRef,
+        initializer: Function,
+      ) {
+        if (decoratorFinishedRef.v) {
+          throw new Error(
+            "attempted to call addInitializer after decoration was finished",
+          );
+        }
+        assertCallable(initializer, "An initializer", "be", true);
+        initializers.push(initializer);
+      }.bind(null, decoratorFinishedRef);
+
+      try {
+        if (isClass) {
+          newValue = dec.call(decThis, Class, ctx);
+        } else {
+          ctx.static = isStatic;
+          ctx.private = isPrivate;
+
+          var get, set;
+          if (!isPrivate && (isField || kind === PROP_KIND.METHOD)) {
+            get = function (target: any) {
+              return target[name];
+            };
+            if (isField) {
+              set = function (target: any, v: any) {
+                target[name] = v;
+              };
+            }
+          } else if (kind === PROP_KIND.METHOD) {
+            // Assert: isPrivate is true.
+            get = function (_this: any) {
+              if (isPrivate) {
```

Comment:
If `isPrivate` is true, why do we need to check it again?

Reply:
This should also move the old code and I will remove this check.